File: litho-core/src/main/java/com/facebook/litho/DrawableMatrix.java
/*
 * Copyright 2014-present Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import android.graphics.Matrix;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.widget.ImageView.ScaleType;
import javax.annotation.Nullable;

/** Static class containing a factory method for creating a matrix to apply to a drawable. */
public final class DrawableMatrix extends Matrix {

  private boolean mShouldClipRect;

  private DrawableMatrix() {}

  /** @return True if this Matrix requires a clipRect() on the bounds of the drawable. */
  public boolean shouldClipRect() {
    return mShouldClipRect;
  }

  /**
   * Create a matrix to be applied to a drawable which scales the drawable according to its scale
   * type.
   *
   * @param d The drawable to create a matrix for
   * @param scaleType A scale type describing how to scale the drawable.
   * @param width The width of the drawable's container.
   * @param height The height of the drawable's container.
   * @return The scale matrix or null if the drawable does not need to be scaled.
   */
  public static @Nullable DrawableMatrix create(
      final Drawable d, ScaleType scaleType, final int width, final int height) {

    if (scaleType == null) {
      scaleType = ScaleType.FIT_CENTER;
    }

    if (d == null) {
      return null;
    }

    final int intrinsicWidth = d.getIntrinsicWidth();
    final int intrinsicHeight = d.getIntrinsicHeight();

    if (intrinsicWidth <= 0
        || intrinsicHeight <= 0
        || ScaleType.FIT_XY == scaleType
        || ScaleType.MATRIX == scaleType) {
      return null;
    }

    if (width == intrinsicWidth && height == intrinsicHeight) {
      // The bitmap fits exactly, no transform needed.
      return null;
    }

    final DrawableMatrix result = new DrawableMatrix();

    if (ScaleType.CENTER == scaleType) {
      // Center bitmap in view, no scaling.
      result.setTranslate(
          FastMath.round((width - intrinsicWidth) * 0.5f),
          FastMath.round((height - intrinsicHeight) * 0.5f));

      result.mShouldClipRect = (intrinsicWidth > width || intrinsicHeight > height);
    } else if (ScaleType.CENTER_CROP == scaleType) {

      final float scale;
      float dx = 0;
      float dy = 0;

      if (intrinsicWidth * height > width * intrinsicHeight) {
        scale = (float) height / (float) intrinsicHeight;
        dx = (width - intrinsicWidth * scale) * 0.5f;
      } else {
        scale = (float) width / (float) intrinsicWidth;
        dy = (height - intrinsicHeight * scale) * 0.5f;
      }

      result.setScale(scale, scale);
      result.postTranslate(FastMath.round(dx), FastMath.round(dy));

      result.mShouldClipRect = true;
    } else if (ScaleType.CENTER_INSIDE == scaleType) {

      final float scale;
      if (intrinsicWidth <= width && intrinsicHeight <= height) {
        scale = 1.0f;
      } else {
        scale =
            Math.min(
                (float) width / (float) intrinsicWidth, (float) height / (float) intrinsicHeight);
      }

      final float dx = FastMath.round((width - intrinsicWidth * scale) * 0.5f);
      final float dy = FastMath.round((height - intrinsicHeight * scale) * 0.5f);

      result.setScale(scale, scale);
      result.postTranslate(dx, dy);
    } else {
      RectF src = new RectF();
      RectF dest = new RectF();

      // Generate the required transform.
      src.set(0, 0, intrinsicWidth, intrinsicHeight);
      dest.set(0, 0, width, height);

      result.setRectToRect(src, dest, scaleTypeToScaleToFit(scaleType));
    }

    return result;
  }

  private static Matrix.ScaleToFit scaleTypeToScaleToFit(ScaleType st) {
    // ScaleToFit enum to their corresponding Matrix.ScaleToFit values
    switch (st) {
      case FIT_XY:
        return Matrix.ScaleToFit.FILL;
      case FIT_START:
        return Matrix.ScaleToFit.START;
      case FIT_CENTER:
        return Matrix.ScaleToFit.CENTER;
      case FIT_END:
        return Matrix.ScaleToFit.END;

      default:
        throw new IllegalArgumentException("Only FIT_... values allowed");
    }
  }
}

File: litho-core/src/main/java/com/facebook/litho/DrawableMatrix.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.graphics.Matrix
import android.graphics.RectF
import android.graphics.drawable.Drawable
import android.widget.ImageView.ScaleType
import com.facebook.rendercore.FastMath
import kotlin.math.min

/** Static class containing a factory method for creating a matrix to apply to a drawable. */
class DrawableMatrix private constructor() : Matrix() {

  private var shouldClipRect = false

  /** @return True if this Matrix requires a clipRect() on the bounds of the drawable. */
  fun shouldClipRect(): Boolean {
    return shouldClipRect
  }

  companion object {

    /**
     * Create a matrix to be applied to a drawable which scales the drawable according to its scale
     * type.
     *
     * @param d The drawable to create a matrix for
     * @param scaleType A scale type describing how to scale the drawable.
     * @param width The width of the drawable's container.
     * @param height The height of the drawable's container.
     * @return The scale matrix or null if the drawable does not need to be scaled.
     */
    @JvmStatic
    fun create(d: Drawable?, scaleType: ScaleType?, width: Int, height: Int): DrawableMatrix? {
      if (d == null) {
        return null
      }

      val scaleType = scaleType ?: ScaleType.FIT_CENTER

      val intrinsicWidth = d.intrinsicWidth
      val intrinsicHeight = d.intrinsicHeight

      if (intrinsicWidth <= 0 ||
          intrinsicHeight <= 0 ||
          ScaleType.FIT_XY == scaleType ||
          ScaleType.MATRIX == scaleType) {
        return null
      }

      if (width == intrinsicWidth && height == intrinsicHeight) {
        // The bitmap fits exactly, no transform needed.
        return null
      }

      val result = DrawableMatrix()

      if (ScaleType.CENTER == scaleType) {
        // Center bitmap in view, no scaling.
        result.setTranslate(
            FastMath.round((width - intrinsicWidth) * 0.5f).toFloat(),
            FastMath.round((height - intrinsicHeight) * 0.5f).toFloat())

        result.shouldClipRect = intrinsicWidth > width || intrinsicHeight > height
      } else if (ScaleType.CENTER_CROP == scaleType) {
        val scale: Float
        var dx = 0f
        var dy = 0f

        if (intrinsicWidth * height > width * intrinsicHeight) {
          scale = height.toFloat() / intrinsicHeight.toFloat()
          dx = (width - intrinsicWidth * scale) * 0.5f
        } else {
          scale = width.toFloat() / intrinsicWidth.toFloat()
          dy = (height - intrinsicHeight * scale) * 0.5f
        }

        result.setScale(scale, scale)
        result.postTranslate(FastMath.round(dx).toFloat(), FastMath.round(dy).toFloat())

        result.shouldClipRect = true
      } else if (ScaleType.CENTER_INSIDE == scaleType) {
        val scale: Float =
            if (intrinsicWidth <= width && intrinsicHeight <= height) {
              1.0f
            } else {
              min(
                  width.toFloat() / intrinsicWidth.toFloat(),
                  height.toFloat() / intrinsicHeight.toFloat())
            }

        val dx = FastMath.round((width - intrinsicWidth * scale) * 0.5f).toFloat()
        val dy = FastMath.round((height - intrinsicHeight * scale) * 0.5f).toFloat()

        result.setScale(scale, scale)
        result.postTranslate(dx, dy)
      } else {
        val src = RectF()
        val dest = RectF()

        // Generate the required transform.
        src[0f, 0f, intrinsicWidth.toFloat()] = intrinsicHeight.toFloat()
        dest[0f, 0f, width.toFloat()] = height.toFloat()

        result.setRectToRect(src, dest, scaleTypeToScaleToFit(scaleType))
      }

      return result
    }

    private fun scaleTypeToScaleToFit(st: ScaleType): ScaleToFit {
      // ScaleToFit enum to their corresponding Matrix.ScaleToFit values
      return when (st) {
        ScaleType.FIT_XY -> ScaleToFit.FILL
        ScaleType.FIT_START -> ScaleToFit.START
        ScaleType.FIT_CENTER -> ScaleToFit.CENTER
        ScaleType.FIT_END -> ScaleToFit.END
        else -> throw IllegalArgumentException("Only FIT_... values allowed")
      }
    }
  }
}
